src/NUnitEngine/nunit.engine.core/Services/ExtensionManager.cs

```diff
@@ -233,14 +254,36 @@ private ExtensionPoint DeduceExtensionPointFromType(TypeReference typeRef)
         }
 
         /// <summary>
-        /// Find candidate extension assemblies starting from a
-        /// given base directory.
+        /// Find candidate extension assemblies starting from a given base directory
+        /// and using the contained '.addins' files to direct the search.
         /// </summary>
-        /// <param name="startDir"></param>
-        private void FindExtensionAssemblies(IDirectory startDir)
+        /// <param name="initialDirectory">Path to the initial directory.</param>
+        private void FindExtensionAssemblies(IDirectory initialDirectory)
+        {
+            // Since no patterns are provided, look for addins files
+            ProcessAddinsFiles(initialDirectory, false);
+        }
+
+        /// <summary>
+        /// Find candidate extension assemblies starting from a given base directory
+        /// and using the provided list of patterns to direct the search using
+        /// a built-in algorithm.
+        /// </summary>
+        /// <param name="initialDirectory">Path to the initial directory.</param>
+        /// <param name="patterns">A list of patterns used to identify potential candidates.</param>
+        private void FindExtensionAssemblies(IDirectory initialDirectory, string[] patterns)
         {
-            // First check the directory itself
-            ProcessAddinsFiles(startDir, false);
+            // Start looking two levels above initial directory
+            var startDir = initialDirectory.Parent.Parent;
```

Comment:
I find this not logical. The user specifies a start directory and then this starts looking two levels higher?
It means it cannot find extension in the current directory.
As the algorithm searches upwards, why not start at the indicated directory?

Reply:
@CharliePoole I have no clear preference, but one of 1, 2 or 3 should be picked.
Should the engine know how a package or an addin is installed?

I know we talked about dropping the dotnet tool, but if that is still there, it would be running from the nuget package cache, addins could be other nuget packages which would be not only be living in parent directories, but at a subdirectory in the nuget cache.

Even if installed in 'program files' I don't know if addins would be installed under the same directory or not.

Reply:
@manfred-brands 
Thanks. I'll pick one of the three before moving to the next step.

I think the engine should only only know how _it_ is installed. Up to now, we have required that packages should be installed in the same way as the engine. This still like a reasonable limitation to me, although there are some areas where it would be convenient to extend it.

In addition, we are only able to load extensions if we know where they are located. I feel that's the biggest weakness of the NUnit extensibility approach and I think this PR can be a step in the right direction. That is, by using an internal algorithm of any kind, we eliminate the need for the _user_ to know where things are located. Over time, that algorithm can be improved.

Your comment about the nuget cache, makes me think of a possible incremental approach to this...
1. Support the nuget cache, i.e. run the engine out of the cache and find extensions in the same place.
2. Do the same for dotnet tools installed locally
3. Repeat for dotnet tools installed in the global cache
4. Extend to allow a globally installed dotnet tool to find extensions installed locally.

BTW, this is how the chocolatey install now works, which is why I tend to prefer it. There is no need for a package that bundles extensions, because users can install and uninstall them as they please.

I suspect that item 1 and possibly 2 of the above incremental steps may already work with the code in this PR, so I'll try to run some tests to find out. I'll see how much I can get done before my course starts next week. If you'd like to try experimenting with any of the above, let me know and I'll focus on the others.

Reply:
Just to clarify:
> Support the nuget cache, i.e. run the engine out of the cache and find extensions in the same place.

I mean that the extensions are other nuget packages.
So if the engine is at: $NUGET_PACKAGES/nunit.engine/version/tool/net8.0/nunit.engine.dll
Extensions could be at: $NUGET_PACKAGES/some.extension/version/lib/net8.0/some.extension.dll

The problem is how do you know there is an extension to look for.
You cannot look for all dlls in $NUGET_PACKAGES as there are a lot.
Do users need to specify, e.g. --addin some.extension and then you find it.

Reply:
We first look for directory names, not for dlls. The package directory has to be named "NUnit.Exception.*" so most directories in $NUGET_PACKAGES would be ignored. If the package is well designed, the tools directory will contain a `.addins` file specifying the exact assemblies to scan. Otherwise, we will scan all dlls under subdirectories "tools" or "tools/*" (to allow for multiply-targeted extensions.

Bear in mind that this PR is not trying to find all extensions, only those in standard format. There are other proposals in issue #488 intended to allow the user to specify additional places to look.